queries/migrations.create_prune_galleries_trigger.sql

```diff
@@ -11,17 +11,50 @@ SELECT g.user_id INTO _user_id FROM galleries g
                                         JOIN wallets w ON g.user_id = w.user_id AND w.id = OLD.wallet_id
     LIMIT 1;
 
--- Cleanup pivot table
+-- Mark all gallery nfts as soft deleted if no longer owned by any user
+WITH unassinged_nfts AS (
+    SELECT gallery_id, nft_id FROM galleries
+                                       JOIN nft_gallery ng on galleries.id = ng.gallery_id
+                                       JOIN nfts n on ng.nft_id = n.id AND n.wallet_id IS NULL
+    WHERE user_id = _user_id
+)
+UPDATE nft_gallery
+    SET deleted_at = NOW()
+FROM unassinged_nfts un
+WHERE nft_gallery.gallery_id = un.gallery_id AND nft_gallery.nft_id = un.nft_id;
+
+-- remove gallery nfts relationship if nft belongs to another user or is deleted
 WITH removed_nfts AS (
     SELECT gallery_id, nft_id FROM galleries
                                        JOIN nft_gallery ng on galleries.id = ng.gallery_id
-                                       JOIN nfts n on ng.nft_id = n.id AND (n.wallet_id IS NULL OR n.wallet_id != OLD.wallet_id OR n.deleted_at IS NOT NULL)
+                                       JOIN nfts n on ng.nft_id = n.id AND (n.wallet_id != OLD.wallet_id OR n.deleted_at IS NOT NULL)
     WHERE user_id = _user_id
 )
 DELETE FROM nft_gallery
     USING removed_nfts rm
 WHERE nft_gallery.gallery_id = rm.gallery_id AND nft_gallery.nft_id = rm.nft_id;
 
+
+-- Mark all galleries with only soft deleted nfts as soft deleted
+WITH soft_deleted_nfts AS (
```

Comment:
Since we know which galleries were affected inside this trigger, we don't need to check all galleries on each invocation.

Instead  we can store the distinct affected gallery ids (see my other comment) via `RETURNING` in a variable to only check those here with something like `WHERE deleted_at IS NOT NULL AND id = ANY (_updated_galleries)`.

Reply:
Tried to apply that suggestion but had some sql errrors that was unable to fix (code and error bwlow) apparently you can't directly collect multiple returned values into an array using RETURNING

What I did instead was to add an extra condition in the query to filter the galleries, [see here](https://github.com/ArdentHQ/dashbrd/pull/176/commits/9ec08a25d46dd78ed45d52476d822134cd5cf92d)

What do you think? 

error:

```
: Invalid text representation: 7 ERROR:  malformed array literal: "9"
DETAIL:  Array value must start with "{" or dimension information.
CONTEXT:  PL/pgSQL function prune_empty_galleries() line 12 at SQL statement (Connection: pgsql, SQL: update "nfts" set "wallet_id" = ?, "updated_at" = 2023-10-04 15:15:40 where "wallet_id" = 21 and exists (select * from "collections" where "nfts"."collection_id" = "collections"."id" and "network_id" = 53 and "address" = 0x0053399124f0cbb46d2cbacd8a89cf0599974963 and "collections"."deleted_at" is null) and "updated_at" < 2023-10-04 15:15:40 and "nfts"."deleted_at" is null)
````

code

```sql
CREATE OR REPLACE FUNCTION prune_empty_galleries() RETURNS TRIGGER AS $_$
DECLARE
    _user_id BIGINT;
    _updated_galleries BIGINT[];
BEGIN
    -- Fetch user id of old wallet
    SELECT g.user_id INTO _user_id FROM galleries g
                                        JOIN wallets w ON g.user_id = w.user_id AND w.id = OLD.wallet_id
    LIMIT 1;

    -- Mark all gallery nfts as soft deleted if no longer owned by any user
    UPDATE nft_gallery
    SET deleted_at = NOW()
    WHERE nft_gallery.nft_id = NEW.id AND NEW.wallet_id IS NULL
    RETURNING nft_gallery.gallery_id INTO _updated_galleries;

    -- remove gallery nfts relationship if nft belongs to another user or is deleted
    -- ... (rest of your function here)

    -- Mark all galleries with only soft deleted nfts as soft deleted
    WITH soft_deleted_nfts AS (
        SELECT gallery_id
        FROM nft_gallery
        WHERE deleted_at IS NOT NULL
        AND gallery_id = ANY (_updated_galleries)
        GROUP BY gallery_id
        HAVING COUNT(*) = COUNT(deleted_at)
    )
    UPDATE galleries
        SET deleted_at = NOW()
    WHERE
        user_id = _user_id
    AND
        EXISTS (
            SELECT 1
            FROM soft_deleted_nfts
            WHERE galleries.id = soft_deleted_nfts.gallery_id
        );

    -- ... (rest of your function here)

    RETURN NEW;
END;
$_$ LANGUAGE plpgsql;
```

Reply:
@alfonsobries Yeah, I think how you did it is also fine. We can keep it as is then. :+1: 

---

Just for completeness, this is how to use it with an array:

```patch
diff --git a/queries/migrations.create_prune_galleries_trigger.sql b/queries/migrations.create_prune_galleries_trigger.sql
index a87b8d48..1285053d 100644
--- a/queries/migrations.create_prune_galleries_trigger.sql
+++ b/queries/migrations.create_prune_galleries_trigger.sql
@@ -5,6 +5,7 @@
 CREATE OR REPLACE FUNCTION prune_empty_galleries() RETURNS TRIGGER AS $_$
 DECLARE
     _user_id BIGINT;
+    _updated_galleries BIGINT[];    
 BEGIN
     -- Fetch user id of old wallet
 SELECT g.user_id INTO _user_id FROM galleries g
@@ -12,9 +13,15 @@ SELECT g.user_id INTO _user_id FROM galleries g
     LIMIT 1;
 
 -- Mark all gallery nfts as soft deleted if no longer owned by any user
-UPDATE nft_gallery
-SET deleted_at = NOW()
-WHERE nft_gallery.nft_id = NEW.id AND NEW.wallet_id IS NULL;
+WITH updated_galleries AS (
+    UPDATE nft_gallery
+    SET deleted_at = NOW()
+    WHERE nft_gallery.nft_id = NEW.id AND NEW.wallet_id IS NULL
+    RETURNING nft_gallery.gallery_id
+)
+SELECT array_agg(gallery_id)
+FROM updated_galleries 
+INTO _updated_galleries;
 
 -- remove gallery nfts relationship if nft belongs to another user or is deleted
 WITH removed_nfts AS (
@@ -27,14 +34,12 @@ DELETE FROM nft_gallery
     USING removed_nfts rm
 WHERE nft_gallery.gallery_id = rm.gallery_id AND nft_gallery.nft_id = rm.nft_id;
 
-
 -- Mark all galleries with only soft deleted nfts as soft deleted
 WITH soft_deleted_nfts AS (
     SELECT gallery_id
     FROM nft_gallery
     WHERE deleted_at IS NOT NULL
-    AND nft_gallery.nft_id = NEW.id
-    AND NEW.wallet_id IS NULL
+    AND nft_gallery.gallery_id = ANY(_updated_galleries)
     GROUP BY gallery_id
     HAVING COUNT(*) = COUNT(deleted_at)
 )

```